src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,62 +80,66 @@ public T runWithAnyNode() {
     }
   }
 
-  private T runWithRetries(final int slot, int attempts, boolean tryRandomNode, JedisRedirectionException redirect) {
-    if (attempts <= 0) {
-      throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
-    }
+  private T runWithRetries(final int slot) {
+    Supplier<Jedis> connectionSupplier = () -> connectionHandler.getConnectionFromSlot(slot);
 
-    Jedis connection = null;
-    try {
+    // If we got one redirection, stick with that and don't try anything else
+    Supplier<Jedis> redirectionSupplier = null;
 
-      if (redirect != null) {
-        connection = this.connectionHandler.getConnectionFromNode(redirect.getTargetNode());
-        if (redirect instanceof JedisAskDataException) {
-          // TODO: Pipeline asking with the original command to make it faster....
-          connection.asking();
-        }
-      } else {
-        if (tryRandomNode) {
-          connection = connectionHandler.getConnection();
+    for (int currentAttempt = 0; currentAttempt < this.maxAttempts; currentAttempt++) {
+      Jedis connection = null;
+      try {
+        if (redirectionSupplier != null) {
+          connection = redirectionSupplier.get();
         } else {
-          connection = connectionHandler.getConnectionFromSlot(slot);
+          connection = connectionSupplier.get();
         }
+        return execute(connection);
+      } catch (JedisNoReachableClusterNodeException e) {
+        throw e;
+      } catch (JedisConnectionException e) {
+        connectionSupplier = handleConnectionProblem(slot, currentAttempt);
+      } catch (JedisRedirectionException e) {
+        redirectionSupplier = handleRedirection(connection, e);
+      } finally {
+        releaseConnection(connection);
       }
+    }
 
-      return execute(connection);
-
-    } catch (JedisNoReachableClusterNodeException jnrcne) {
-      throw jnrcne;
-    } catch (JedisConnectionException jce) {
-      // release current connection before recursion
-      releaseConnection(connection);
-      connection = null;
-
-      if (attempts <= 1) {
-        //We need this because if node is not reachable anymore - we need to finally initiate slots
-        //renewing, or we can stuck with cluster state without one node in opposite case.
-        //But now if maxAttempts = [1 or 2] we will do it too often.
-        //TODO make tracking of successful/unsuccessful operations for node - do renewing only
-        //if there were no successful responses from this node last few seconds
-        this.connectionHandler.renewSlotCache();
-      }
+    throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
```

Comment:
Minor enhancement: can we track the last exception and add it as the cause exception here?

```suggestion
    throw new JedisClusterMaxAttemptsException("No more cluster attempts left.", lastException);
```